xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -0,0 +1,128 @@
+
+
+
+  /**
+   * The init function of the logic script
+   * For each livedata element on the page, returns its corresponding data / API
+   * If the data does not exists yet, create it from the element
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata component
+   */
+  var init = function (element) {
+
+    if (!instancesMap.has(element)) {
+      // create a new logic object associated to the element
+      var logic = new Logic(element);
+      instancesMap.set(element, logic);
+
+      logic.changeLayout();
+    }
+
+    return instancesMap.get(element);
+  };
+
+
+  /**
+   * Class for a logic element
+   * Contains the Livedata data object and methods to mutate it
+   * Can be used in the layouts to display the data, and call its API
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata
+   */
+  var Logic = function (element) {
+    this.element = element;
+    this.data = JSON.parse(element.getAttribute("data-data") || "{}");
+    element.removeAttribute("data-data");
+    this.layouts = {};
+  };
+
+
+  /**
+   * Change layout to specified one, or default one if none specified
+   * @param {String} layoutId The id of the layout to load with requireJS
+   * @returns {Object} A jquery promise
+   */
+  Logic.prototype.changeLayout = function (layoutId) {
+    var self = this;
+    var defer = $.Deferred();
+
+    layoutId = layoutId || this.data.meta.defaultLayout;
+    if (layoutId === this.data.query.currentLayout) { return defer.reject(); }
+    if (!this.data.meta.layoutDescriptors[layoutId]) { return defer.reject(); }
+    if (this.data.meta.layouts.indexOf(layoutId) === -1) { return defer.reject(); }
+
+    // load layout based on it's filename
+    require([BASE_PATH + "layouts/" + this.data.meta.layoutDescriptors[layoutId].file],
+      // load success
+      function (Layout) {
+        // remove current layout from the page
+        if (self.data.query.currentLayout && self.layouts[self.data.query.currentLayout]) {
+          self.element.removeChild(self.layouts[self.data.query.currentLayout]);
+        }
+        // add layout element in loaded layouts list if not already loaded on the page
+        if (!self.layouts[layoutId]) {
+          self.layouts[layoutId] = new Layout(self.element);
+        }
+        // add new layout to the page
+        self.element.appendChild(self.layouts[layoutId]);
+        self.data.query.currentLayout = layoutId;
+        // dispatch events
+        var event = new CustomEvent("xwiki:livedata:layoutChange", {
+          layoutId: layoutId,
+          livedata: self,
+        });
+        self.element.dispatchEvent(event);
+        defer.resolve();
+      },
+
+      // load failure
+      function (err) {
+        // try to load default layout instead
+        if (layoutId !== self.data.meta.defaultLayout) {
+          return self.changeLayout(self.data.meta.defaultLayout);
+        }
+        else {
```

Comment:
Should be on the same line as ``}``, at least that's our current code style.

Reply:
oh noooo, I hate this code style I find it very unreadable imo :(
But if it's the xwiki code style, I'll stick to it of course.

Reply:
There is no definitive code style for JavaScript but we try to use as much as possible the same code style as on Java (server side). One exception is the indentation (2 spaces on JavaScript and 4 on Java) but there are others. I don't see why we wouldn't use the same code style for ``if / then / else`` in Java and JavaScript since it's the same syntax. Now, these rules are not set in stone but the process of changing them is not easy as it requires the vote of the majority of committers.